opengever/maintenance/scripts/fix_journal_message_mappings.py

```diff
@@ -0,0 +1,98 @@
+from ftw.journal.config import JOURNAL_ENTRIES_ANNOTATIONS_KEY
+from ftw.journal.interfaces import IAnnotationsJournalizable
+from ftw.upgrade import ProgressLogger
+from opengever.maintenance.debughelpers import setup_app
+from opengever.maintenance.debughelpers import setup_option_parser
+from opengever.maintenance.debughelpers import setup_plone
+from plone import api
+from zope.annotation.interfaces import IAnnotations
+from zope.i18nmessageid import Message
+import logging
+import sys
+import transaction
+
+
+logger = logging.getLogger('opengever.maintenance')
+handler = logging.StreamHandler(stream=sys.stdout)
+logging.root.addHandler(handler)
+logging.root.setLevel(logging.INFO)
+
+MAPPING_KEYS = ['title', 'repository']
+
+
+class JournalMessageFixer(object):
+    """Fixes encoding for zope.i18nmessageid.Message mapping values, for
+    journal entries of objects which has been created with an OGGBundle import.
+
+    See https://github.com/4teamwork/opengever.core/pull/3369 for more details.
+    """
+
+    def __init__(self, portal, options):
+        self.portal = portal
+        self.options = options
+        self.counter = 0
+
+    def get_journal_entries(self, obj):
+        """Returns all journal entries for the given object.
+        """
+        if IAnnotationsJournalizable.providedBy(obj):
+            annotations = IAnnotations(obj)
+            return annotations.get(JOURNAL_ENTRIES_ANNOTATIONS_KEY, [])
+
+        return []
+
+    def fix_entries(self, obj):
+        for entry in self.get_journal_entries(obj):
+            message = entry['action'].get('title')
+            if message and isinstance(message, Message) and message.mapping:
+                if self.check_message(message):
+                    self.fix_message(entry, message)
+
+    def check_message(self, message):
+        for key in MAPPING_KEYS:
+            if key in message.mapping:
+                value = message.mapping[key]
+                if not isinstance(value, unicode):
+                    return True
+        return False
+
+    def fix_message(self, entry, old_message):
+        """Replaces the existing Messsage object with a new one,
+        with mapping values in the correct encoding (unicode).
+        """
+        mapping = old_message.mapping
+        for key in mapping.keys():
+            if key in MAPPING_KEYS:
+                value = mapping[key]
+                if not isinstance(value, unicode):
+                    mapping[key] = value.decode('utf-8')
+
+        entry['action']['title'] = Message(
```

Comment:
i don't understand (from the code) why this is necessary, wouldn't it be enough to just replace the mapping values? could you add a short comment?

Reply:
The docstring of the method describes what it does. I now also add a line why it's necessary.